presto-hive/src/test/java/com/facebook/presto/hive/parquet/AbstractTestParquetReader.java

```diff
@@ -103,6 +114,281 @@ public void setUp()
     }
 
     @Test
+    public void testArray()
+            throws Exception
+    {
+        Iterable<List<Integer>> values = createTestArray(limit(cycle(Arrays.asList(1, null, 3, 5, null, null, null, 7, 11, null, 13, 17)), 30_000));
+        tester.testRoundTrip(getStandardListObjectInspector(javaIntObjectInspector), values, values, new ArrayType(INTEGER));
+    }
+
+    @Test
+    public void testArrayAllEmpty()
+            throws Exception
+    {
+        Iterable<List<Integer>> values = limit(cycle(Collections.singletonList(Collections.emptyList())), 30_000);
+        tester.testRoundTrip(getStandardListObjectInspector(javaIntObjectInspector), values, values, new ArrayType(INTEGER));
+    }
+
+    @Test
+    public void testNestedArrays()
+            throws Exception
+    {
+        Iterable<List<List<Integer>>> values = createTestArray(createTestArray(limit(cycle(ImmutableList.of(1, 3, 5, 7, 11, 13, 17)), 30_000)));
+        tester.testRoundTrip(getStandardListObjectInspector(getStandardListObjectInspector(javaIntObjectInspector)), values, values, new ArrayType(new ArrayType(INTEGER)));
+    }
+
+    @Test
+    public void testArrayOfArrayOfArray()
+            throws Exception
+    {
+        Iterable<List<List<List<Integer>>>> values = createTestArray(createTestArray(createTestArray(limit(cycle(Arrays.asList(1, null, 3, null, 5, null, 7, null, null, null, 11, null, 13)), 130_000))));
+        tester.testRoundTrip(
+                getStandardListObjectInspector(getStandardListObjectInspector(getStandardListObjectInspector(javaIntObjectInspector))),
+                values, values, new ArrayType(new ArrayType(new ArrayType(INTEGER))));
+    }
+
+    @Test
+    public void testArrayOfStruct()
+            throws Exception
+    {
+        List<List> structs = createTestStruct(transform(intsBetween(0, 31_234), Object::toString), longsBetween(0, 31_234));
+        Iterable<List<List>> values = createTestArray(structs);
+        List<String> structFieldNames = Arrays.asList("stringParam", "longParam");
+        Type structType = new RowType(ImmutableList.of(VARCHAR, BIGINT), Optional.of(structFieldNames));
+        tester.testRoundTrip(
+                getStandardListObjectInspector(getStandardStructObjectInspector(structFieldNames, Arrays.asList(javaStringObjectInspector, javaLongObjectInspector))),
+                values, values, new ArrayType(structType));
+    }
+
+    @Test
+    public void testArrayOfArrayOfStruct()
```

Comment:
added testStructOfNullableMapBetweenNonNullFields and testStructOfNullableArrayBetweenNonNullFields. Also changed all other test so that nested structs/arrays/maps have null values, not only contain inner null elements.